crates/bevy_hierarchy/src/one_to_many/many.rs

```diff
@@ -0,0 +1,308 @@
+#[cfg(feature = "reflect")]
+use bevy_ecs::reflect::{
+    ReflectComponent, ReflectFromWorld, ReflectMapEntities, ReflectVisitEntities,
+    ReflectVisitEntitiesMut,
+};
+use bevy_ecs::{
+    component::{Component, ComponentId},
+    entity::{Entity, VisitEntitiesMut},
+    event::Events,
+    world::{DeferredWorld, World},
+};
+use core::{fmt::Debug, marker::PhantomData, ops::Deref};
+use smallvec::SmallVec;
+
+use super::{ManyToOne, OneToManyEvent};
+
+/// Represents one half of a one-to-many relationship between an [`Entity`] and some number of other [entities](Entity).
+///
+/// The type of relationship is denoted by the parameter `R`.
+#[derive(Component)]
+#[component(
+    on_insert = Self::associate,
+    on_replace = Self::disassociate,
+    on_remove = Self::disassociate
```

Comment:
I may be absolutely, totally, wrong here as this is the very first time I find out that there even is an `on_replace` hook, and I've never even used hooks before, but...

Documentation says `on_replace` is also run when the component gets removed from the object, so why do we need `on_remove` as well?

I think it works because calling disassociate twice is safe. And I didn't add comments on all of the `on_remove` spots so all of them need to be changed if this is.

Reply:
Correct! I've explicitly made these methods safe for being run multiple times to account for this. When I originally wrote this I remembered finding an instance where `on_replace` was _not_ called after `on_remove`, but I'll double-check that now.

Reply:
If you do find it, it would probably be a good idea to add a test case which fails if `on_remove` is not given. Then if it is fixed, it's easy to see if the test case passes too.

Reply:
Ok found it [here](https://github.com/bevyengine/bevy/blob/d15d901cab763a1619f4dffbfd0418618daebe7f/crates/bevy_ecs/src/bundle.rs#L904). The `BundleInserter` `insert` method will trigger the `on_replace` hook without first calling any `on_remove` hooks (if applicable). Because of that, you _must_ have hooks for both `on_remove` _and_ `on_replace` to capture all cases of a `Component` being taken off of an `Entity`.

Reply:
Although your original question holds, since if `on_remove` is called,  `on_replace` is always called, so I _think_ you can omit the `on_remove` hook. I'll test that now...

Reply:
Yeah you're right, removing the `on_remove` for the sake of simplicity. Good catch!

Reply:
Yeah, I believe the way the hooks work is:
- on_insert gets called every time a component is added by any means
- on_replace gets called every time a component is removed by any means
- on_add gets called only if the component didn't exist yet on the entity
- on_remove gets called only if the component will no longer exist on the entity

So on_replace indeed catches all "removals".

Good to hear there weren't any bugs uncovered by this, as `on_replace` seems to be quite new.